File: Hilos/PlanificadorTareas.py
import time
import heapq

class PlanificadorTareas:
    def __init__(self, nombre, tiempo_ejecucion):
        self.nombre = nombre
        self.tiempo_ejecucion = tiempo_ejecucion
        self.tiempo_restante = tiempo_ejecucion
        self.tiempo_inicio = 0
        self.tiempo_fin = 0

def ejecutarFCFS(tareas):
    tiempo_actual = 0
    for tarea in tareas:
        tarea.tiempo_inicio = tiempo_actual
        tarea.tiempo_fin = tiempo_actual + tarea.tiempo_ejecucion
        tiempo_actual = tarea.tiempo_fin
        print(f"{tarea.nombre} se inició a los {tarea.tiempo_inicio} y terminó a los {tarea.tiempo_fin}")
        time.sleep(0.5)

def ejecutarRR(tareas, quantum):
    tiempo_actual = 0
    while tareas:
        for tarea in tareas[:]:
            if tarea.tiempo_restante > quantum:
                print(f"{tarea.nombre} se ejecuta desde {tiempo_actual} hasta {tiempo_actual + quantum}")
                tarea.tiempo_restante -= quantum
                tiempo_actual += quantum
                time.sleep(1)
            else:
                print(f"{tarea.nombre} se ejecuta desde {tiempo_actual} hasta {tiempo_actual + tarea.tiempo_restante}")
                tiempo_actual += tarea.tiempo_restante
                time.sleep(1)
                tareas.remove(tarea)

def ejecutarSRT(tareas):
    cola_tareas = [(t.tiempo_restante, t) for t in tareas]
    heapq.heapify(cola_tareas)
    
    tiempo_actual = 0
    while cola_tareas:
        _, tarea = heapq.heappop(cola_tareas)
        if tarea.tiempo_restante > 0:
            print(f"{tarea.nombre} se ejecuta desde {tiempo_actual} hasta {tiempo_actual + tarea.tiempo_restante}")
            tiempo_actual += tarea.tiempo_restante
            tarea.tiempo_restante = 0

tareas = [
    PlanificadorTareas("Tarea 1", 5),
    PlanificadorTareas("Tarea 2", 8),
    PlanificadorTareas("Tarea 3", 3),
    PlanificadorTareas("Tarea 4", 9),
    PlanificadorTareas("Tarea 5", 6)
]

print("Selecciona el algoritmo de planificación")
print("1. FCFS (First-come, First-served)")
print("2. RR (Round Robin)")
print("3. SRT (Shortest Remaining Time)")

opcion = int(input())
if opcion == 1:
    ejecutarFCFS(tareas)
    ejecutarFCFS(tareas)
elif opcion == 2:
    ejecutarRR(tareas, 2)
    ejecutarRR(tareas, 2)
elif opcion == 3:
    ejecutarSRT(tareas)
    ejecutarSRT(tareas)

print("Fin del planificador de tareas")

File: Hilos Java/Hilos/PlanificadorTareas.java
import java.util.*;

class Tarea {
    String nombre;
    int tiempoEjecucion;
    int tiempoRestante;
    int tiempoInicio;
    int tiempoFin;

    public Tarea(String nombre, int tiempoEjecucion) {
        this.nombre = nombre;
        this.tiempoEjecucion = tiempoEjecucion;
        this.tiempoRestante = tiempoEjecucion;

    }

}

public class PlanificadorTareas {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Selecciona el algoritmo de planificación");
        System.out.println("1. FCFS (First-come, First-Served)");
        System.out.println("2. RR (Round Robin)");
        System.out.println("3. SRT (Shortest Remaining Time)");
        int opcion = scanner.nextInt();

        List<Tarea> tareas = new ArrayList<>(Arrays.asList(
                new Tarea("Tarea 1", 5),
                new Tarea("Tarea 2", 8),
                new Tarea("Tarea 3", 3),
                new Tarea("Tarea 4", 9),
                new Tarea("Tarea 5", 6)));

        switch (opcion) {
            case 1:
                ejecutarFCFS(tareas);
                break;
            case 2:
                ejecutarRR(tareas, 2);
                break;
            case 3:
                ejecutarSRT(tareas);
                break;

            default:
                break;
        }
    }

    private static void ejecutarFCFS(List<Tarea> tareas) {
        int tiempoActual = 0;
        for (Tarea tarea : tareas) {
            tarea.tiempoInicio = tiempoActual;
            tarea.tiempoFin = tiempoActual + tarea.tiempoEjecucion;
            tiempoActual = tarea.tiempoFin;
        }
        for (Tarea tarea : tareas) {
            System.out.println(
                    tarea.nombre + "se inicio a los" + tarea.tiempoInicio + "y terminó a los " + tarea.tiempoFin);
            try {
                Thread.sleep(500);
            } catch (InterruptedException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }
    }

    private static void ejecutarRR(List<Tarea> tareas, int quantum) {
        int tiempoActual = 0;
        while (!tareas.isEmpty()) {
            Iterator<Tarea> iterator = tareas.iterator();
            while (iterator.hasNext()) {
                Tarea tarea = iterator.next();
                if (tarea.tiempoRestante > quantum) {
                    System.out.println(
                            tarea.nombre + " se ejecuta desde " + tiempoActual + " hasta " + (tiempoActual + quantum));
                    tarea.tiempoRestante -= quantum;
                    tiempoActual += quantum;
                    try {
                        Thread.sleep(1000);
                    } catch (InterruptedException e) {
                        // TODO Auto-generated catch block
                        e.printStackTrace();
                    }
                } else {
                    System.out.println(tarea.nombre + " se ejecuta desde " + tiempoActual + " hasta "
                            + (tiempoActual + tarea.tiempoRestante));
                    tiempoActual += tarea.tiempoRestante;
                    try {
                        Thread.sleep(1000);
                    } catch (InterruptedException e) {
                        // TODO Auto-generated catch block
                        e.printStackTrace();
                    }
                    iterator.remove();
                }
            }

        }
    }

    private static void ejecutarSRT(List<Tarea> tareas) {
        PriorityQueue<Tarea> colaTareas = new PriorityQueue<>(Comparator.comparing(t -> t.tiempoRestante));
        colaTareas.addAll(tareas);

        int tiempoActual = 0;
        while (!colaTareas.isEmpty()) {
            Tarea tarea = colaTareas.poll();
            if (tarea.tiempoRestante > 0) {
                System.out.println(tarea.nombre + " se ejecuta desde " + tiempoActual + " hasta "
                        + (tiempoActual + tarea.tiempoRestante));
                tiempoActual += tarea.tiempoRestante;
                tarea.tiempoRestante = 0;

            }

        }

    }
}
